File: Java/Language/InterfaceTest1/src/Account.java
package edu.met.banking;

//a class declared with abstract modifier cannot be instantiated
public abstract class Account {

	long id;
	protected double balance;

	public long getId() { return id; }

	public double getBalance() { return balance; }

	//a method declared with abstract modifier must be
	//overridden in a sub-class to support instantiation
	public abstract void deposit(double amount);

	public abstract void withdraw(double amount) throws InsufficientFundsException;

	//a method declared with final modifier cannot be overridden
	//in a sub-class (JVM can skip dynamic binding for such method)
	public final void transfer(double amount, Account that) throws InsufficientFundsException {
		if(this == that)
			throw new IllegalTransferException();
		this.withdraw(amount);
		that.deposit(amount);
	}
}

File: dotnet/Language/Inheritance/InterfaceTest1/BankLib/Account.cs
namespace Met.Banking;

//a class defined with abstract modifier cannot be
//activated, it can only be used as a base class
public abstract class Account
{
    public long Id { get; internal set; }

    public double Balance { get; protected set; }

    //an instance method defined with abstract modifier
    //is pure and it must be overridden in a non-abstract
    //derived class
    public abstract void Deposit(double amount);

    public abstract void Withdraw(double amount);

    public bool Transfer(double amount, Account that)
    {
        if(ReferenceEquals(this, that))
            return false;
        this.Withdraw(amount);
        that.Deposit(amount);
        return true;
    }
}
